core/http-auth-aws-event-stream/src/main/java/software/amazon/awssdk/http/auth/aws/eventstream/internal/SigV4DataFramePublisher.java

```diff
@@ -0,0 +1,233 @@
+
+    public SigV4DataFramePublisher(Publisher<ByteBuffer> publisher,
+                                   AwsCredentialsIdentity credentials,
+                                   CredentialScope credentialScope,
+                                   String signature,
+                                   Clock signingClock) {
+
+        // Adapt the publisher with a trailing-empty frame publisher
+        Publisher<ByteBuffer> trailingPublisher = new TrailingDataFramePublisher(publisher);
+
+        // Map publisher with signing function
+        this.sigv4Publisher = subscriber -> {
+            Subscriber<ByteBuffer> adaptedSubscriber = MappingSubscriber.create(subscriber, getDataFrameSigner(credentials,
+                                                                                                               credentialScope,
+                                                                                                               signature,
+                                                                                                               signingClock
+                                                                                )
+            );
+            trailingPublisher.subscribe(adaptedSubscriber);
+        };
+    }
+
+    private static Function<ByteBuffer, ByteBuffer> getDataFrameSigner(AwsCredentialsIdentity credentials,
+                                                                       CredentialScope credentialScope,
+                                                                       String signature,
+                                                                       Clock signingClock) {
+        return new Function<ByteBuffer, ByteBuffer>() {
+
+            /**
+             * Initiate rolling signature with an initial signature
+             */
+            String priorSignature = signature;
+
+            @Override
+            public ByteBuffer apply(ByteBuffer byteBuffer) {
+                /**
+                 * Signing Date
+                 */
+                Map<String, HeaderValue> eventHeaders = new HashMap<>();
+                Instant signingInstant = signingClock.instant();
+                eventHeaders.put(":date", HeaderValue.fromTimestamp(signingInstant));
+
+                /**
+                 * Derive Signing Key - since a stream of events could be over a period of time, we should update
+                 * the credential scope with the new instant every time the data-frame signer is called
+                 */
+                CredentialScope updatedCredentialScope = new CredentialScope(credentialScope.getRegion(),
+                                                                             credentialScope.getService(), signingInstant);
+                byte[] signingKey = deriveSigningKey(credentials, updatedCredentialScope);
+
+                /**
+                 * Calculate rolling signature
+                 */
+                byte[] payload = byteBuffer.array();
```

Comment:
This feels wrong. What if the provided byte buffer doesn't have an array? What if the sender only wants us to read a portion of the data?

The "safe" and "right" thing is to do a `byte[] payload = copyBytesFrom(byteBuffer)`. Can we test doing that? The upstream might need to be updated to ensure the byte buffer position is at the start and the limit is at the end.

Reply:
Yeah, that's a good point. This was copied over, but I'll make the change.